src/lib/Lower.hs

```diff
@@ -0,0 +1,130 @@
+    _ -> traverseExprDefault expr
+
+type Dest = Atom
+
+traverseFor :: Emits o => Maybe (Dest o) -> ForAnn -> Atom i -> LamExpr i -> LowerM i o (Expr o)
+traverseFor maybeDest dir ixTy lam@(LamExpr (LamBinder ib ty arr eff) body) = do
+  initDest <- case maybeDest of
+    Just  d -> return d
+    Nothing -> emitOp . AllocDest =<< getTypeSubst (Hof $ For dir ixTy (Lam lam))
+  destTy <- getType initDest
+  ty' <- substM ty
+  ixTy' <- substM ixTy
+  eff' <- substM $ ignoreHoistFailure $ hoist ib eff
+  body' <- buildLam noHint arr (PairTy ty' destTy) eff' \b' -> do
+    (i, dest) <- fromPair $ Var b'
+    idest <- emitOp $ IndexRef dest i
+    extendSubst (ib @> SubstVal i) $ traverseBlockWithDest idest body
+    return UnitVal
+  let seqHof = Hof $ Seq dir ixTy' initDest body'
+  case maybeDest of
+    Just _  -> return seqHof
+    Nothing -> Op . Freeze . Var <$> emit seqHof
+
+-- Destination-passing traversals
+
+type AtomNameMap (i'::S) (o::S) = M.Map (AtomName i') (Dest o)
+type DestAssignment (i'::S) (o::S) = AtomNameMap i' o
+
+lookupDest :: DestAssignment i' o -> AtomName i' -> Maybe (Dest o)
+lookupDest = flip M.lookup
+
+decomposeDest :: Emits o => Dest o -> Atom i' -> LowerM i o (Maybe (DestAssignment i' o))
+decomposeDest dest = \case
+  Var v -> return $ Just $ M.singleton v dest
+  _ -> return Nothing
+
+traverseBlockWithDest :: Emits o => Dest o -> Block i -> LowerM i o ()
+traverseBlockWithDest dest (Block _ decls ans) = do
+  decomposeDest dest ans >>= \case
+    Nothing -> do
+      ans' <- traverseDeclNest decls $ traverseAtom ans
+      void $ emitOp $ Place dest ans'
+    Just destMap -> do
+      s <- getSubst
+      case isDistinctNest decls of
+        Nothing -> error "Non-distinct decls?"
+        Just DistinctBetween -> traverseDeclNestWithDestS destMap s decls
+        -- Note that we ignore ans! Its components are written inplace through destMap.
+
+traverseDeclNestWithDestS
+  :: forall i i' l o. (Emits o, DistinctBetween l i')
+  => DestAssignment i' o -> Subst AtomSubstVal l o -> Nest Decl l i' -> LowerM i o ()
+traverseDeclNestWithDestS destMap s = \case
+  Empty -> return ()
+  Nest (Let b (DeclBinding ann _ expr)) rest -> do
+    DistinctBetween <- return $ withExtEvidence rest $ shortenBetween @i' b
+    let maybeDest = lookupDest destMap $ sinkBetween $ binderName b
+    expr' <- withSubst s $ traverseExprWithDest maybeDest expr
+    v <- emitDecl (getNameHint b) ann expr'
+    traverseDeclNestWithDestS destMap (s <>> (b @> Rename v)) rest
+
+traverseExprWithDest :: Emits o => Maybe (Dest o) -> Expr i -> LowerM i o (Expr o)
+traverseExprWithDest dest expr = case expr of
+  Hof (For dir ixTy (Lam body)) -> traverseFor dest dir ixTy body
+  _ -> do
+    expr' <- traverseExprDefault expr
+    case dest of
+      Nothing -> return expr'
+      Just d  -> do
+        ans <- Var <$> emit expr'
+        void $ emitOp $ Place d ans
+        return $ Atom ans
```

Comment:
I am confused about the intended relationship between writing a value to a dest and being able to reference that value through its atom as well.  Shouldn't the dest be the source of truth, such that we emit a reconstruction from it here?  Or am I missing something?

Reply:
The dest is where the result of the whole block will be written, but it doesn't have to be the only use site of that value. Consider
```
v0 = x + 2
v1 = v0 * 2
(v0, v1)
```
if we split the destination for a pair into `v0` and `v1` we can't replace `v0 = x + 2` with a `place`, because it would make it unavailable for the computation of `v1`. What you see here effectively lifts the `place` ops from the end of the block to where the equations are. That is, instead of
```
v0 = ...
v1 = ...
place d0 v0
place d1 v1
```
we'll end up with
```
v0 = ...
place d0 v0
v1 = ...
place d1 v1
```